service/controller/resource/workermigration/create.go

```diff
@@ -0,0 +1,496 @@
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted built-in workers' VMSS")
+
+	return nil
+}
+
+func (r *Resource) allDrainerConfigsWithDrainedState(ctx context.Context, cr providerv1alpha1.AzureConfig) (bool, error) {
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cr),
+	}
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	for _, dc := range dcList.Items {
+		if !dc.Status.HasDrainedCondition() {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *Resource) deleteTimedOutDrainerConfigs(ctx context.Context, cr providerv1alpha1.AzureConfig) error {
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cr),
+	}
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, dc := range dcList.Items {
+		if dc.Status.HasTimeoutCondition() {
+			err = r.ctrlClient.Delete(ctx, &dc)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) deleteDrainerConfigs(ctx context.Context, cr providerv1alpha1.AzureConfig) error {
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cr),
+	}
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, dc := range dcList.Items {
+		err = r.ctrlClient.Delete(ctx, &dc)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) ensureAzureMachinePoolExists(ctx context.Context, cr providerv1alpha1.AzureConfig, vmSize string) (expcapzv1alpha3.AzureMachinePool, error) {
+	var azureMachinePool expcapzv1alpha3.AzureMachinePool
+
+	nsName := types.NamespacedName{
+		Name:      cr.GetName(),
+		Namespace: cr.GetNamespace(),
+	}
+	err := r.ctrlClient.Get(ctx, nsName, &azureMachinePool)
+	if err == nil {
+		// NOTE: Success return when AzureMachinePool CR already exists.
+		return azureMachinePool, nil
+	} else if errors.IsNotFound(err) {
+		// This is ok. CR gets created in a bit.
+	} else if err != nil {
+		return expcapzv1alpha3.AzureMachinePool{}, microerror.Mask(err)
+	}
+
+	var dockerDiskSizeGB int
+	var kubeletDiskSizeGB int
+	{
+		if key.WorkerCount(cr) > 0 {
+			dockerDiskSizeGB = cr.Spec.Azure.Workers[0].DockerVolumeSizeGB
+			kubeletDiskSizeGB = cr.Spec.Azure.Workers[0].KubeletVolumeSizeGB
+		}
+
+		if dockerDiskSizeGB <= 0 {
+			dockerDiskSizeGB = 100
+		}
+
+		if kubeletDiskSizeGB <= 0 {
+			kubeletDiskSizeGB = 100
+		}
+	}
+
+	// CR didn't exist so it's created here.
+	azureMachinePool = expcapzv1alpha3.AzureMachinePool{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AzureMachinePool",
+			APIVersion: "exp.infrastructure.cluster.x-k8s.io/v1alpha3",
```

Comment:
Are there properties in the `azureMachinePool` variable that we can use instead of _hardcoding_ this here? I mean something like
```
Kind: azureMachinePool.Kind,
APIVersion: azureMachinePool.APIVersion,
```

That way, if version changes, we don't need to manually change it.